File: Data structures/Problems/reverse_string.ts
// reverse a string
var str_1:string = "Hello";
var str_2:string = "";
var str_3:string = "k";
var str_4:string = "hi there";

const reverseString: (s: string) => string = function(
    s: string
): string {
    let left:number = 0
    let right:number = s.length - 1
    let sArray:Array<string> = s.split("")
    while (right > left){
        let letter:string = sArray[right]
        sArray[right] = sArray[left]
        sArray[left] = letter
        left += 1
        right -= 1
    }
    let newS:string = sArray.join("")
    return newS;
}

console.log(reverseString(str_1));
console.log(reverseString(str_2));
console.log(reverseString(str_3));
console.log(reverseString(str_4));

File: Data structures/Problems/reverse_string.py
# Reverse a string
str_1 = "Hello"
str_2 = ""
str_3 = "k"
str_4 = "hi there"

# O(log N)

def reverse_string(s: str) -> str:
    left = 0
    right = len(s) - 1
    str_list = list(s)
    while right > left:

        interm = str_list[right]
        str_list[right] = str_list[left]
        str_list[left] = interm
        left += 1
        right -= 1

    return "".join(str_list)

print(reverse_string(str_1))
print(reverse_string(str_2))
print(reverse_string(str_3))
print(reverse_string(str_4))
